lib/lte_link_control/lte_lc_helpers.c

```diff
@@ -567,3 +622,241 @@ int parse_xt3412(const char *at_response, uint64_t *time)
 	at_params_list_free(&resp_list);
 	return err;
 }
+
+uint32_t neighborcell_count_get(const char *at_response)
+{
+	uint32_t comma_count, ncell_elements, ncell_count;
+
+	if (at_response == NULL) {
+		LOG_ERR("at_response is NULL, can't get cell count");
+		return 0;
+	}
+
+	comma_count = get_char_frequency(at_response, ',');
+	if (comma_count < AT_NCELLMEAS_PRE_NCELLS_PARAMS_COUNT) {
+		return 0;
+	}
+
+	/* Add one, as there's no comma after the last element. */
+	ncell_elements = comma_count - (AT_NCELLMEAS_PRE_NCELLS_PARAMS_COUNT - 1) + 1;
+	ncell_count = ncell_elements / AT_NCELLMEAS_N_PARAMS_COUNT;
+
+	return ncell_count;
+}
+
+/* Parse NCELLMEAS notification and put information into struct lte_lc_cells_info.
+ *
+ * Returns 0 on successful cell measurements and population of struct, 1 on
+ * measurement failure, -E2BIG if not all cells were parsed due to memory
+ * limitations, otherwise negative error code.
+ */
+int parse_ncellmeas(const char *at_response, struct lte_lc_cells_info *cells)
+{
+	int err, status, tmp, len;
+	struct at_param_list resp_list;
+	char  response_prefix[sizeof(AT_NCELLMEAS_RESPONSE_PREFIX)] = {0};
+	size_t response_prefix_len = sizeof(response_prefix);
+	char tmp_str[7];
+	bool incomplete = false;
+	/* Count the actual numbers of parameters in the AT response before
+	 * allocating heap for it. This may save quite a bit of heap as the
+	 * worst case scenario is 96 elements.
+	 * 3 is added to account for the parameters that do not have a trailng
+	 * comma.
+	 */
+	size_t param_count = get_char_frequency(at_response, ',') + 3;
+
+	err = at_params_list_init(&resp_list, param_count);
+	if (err) {
+		LOG_ERR("Could not init AT params list, error: %d", err);
+		return err;
+	}
+
+	err = at_parser_params_from_str(at_response,
+					NULL,
+					&resp_list);
+	if (err && err != -E2BIG) {
+		LOG_ERR("Could not parse AT%%XNCELLMEAS  response, error: %d", err);
+		goto clean_exit;
+	} else if (err == -E2BIG) {
+		incomplete = true;
+	}
+
+	err = at_params_string_get(&resp_list,
+				   AT_RESPONSE_PREFIX_INDEX,
+				   response_prefix,
+				   &response_prefix_len);
+	if (err) {
+		LOG_ERR("Could not get response prefix, error: %d", err);
+		goto clean_exit;
+	}
+
+	if (!response_is_valid(response_prefix, response_prefix_len,
+			       AT_NCELLMEAS_RESPONSE_PREFIX)) {
+		/* The unsolicited response is not a CEREG response, ignore it.
+		 */
+		LOG_DBG("Not a valid XNCELLMEAS response");
+		goto clean_exit;
+	}
+
+	/* Status code. */
+	err = at_params_int_get(&resp_list, AT_NCELLMEAS_STATUS_INDEX, &status);
+	if (err) {
+		goto clean_exit;
+	}
+
+	if (status != AT_NCELLMEAS_STATUS_VALUE_SUCCESS) {
+		return 1;
```

Comment:
Memory leak? it should jump to `clean_exit` to release the list I guess?

Reply:
Yeah, that's right.. I even tested for memory leaks this time around, but that didn't help much apparently. Fixed now and added a test case that would have shown the leak.